docs/guidelines.md

```diff
@@ -0,0 +1,285 @@
+
+impl core::Store for Store {}
+
+pub fn create_store() -> impl core::Store {
+  Store
+}
+```
+
+Now the `data` module depends on the `core` module and the core module does not even have
+a dependency on the `data` module at all. Furthermore, the interface that the `core` module
+needs is defined in the `Store` trait, directly in the `core` module.
+
+This idea of using traits is very neat and practical if you want to mock. If you don't want
+to mock using modules is more pragmatic, but always keep in mind the understanding that `core`
+dictates the interface that `data` exposes.
+
+To summarize, the above approach of separating core logic allows you to put an accent on
+the code that is most important and leave boilerplate decisions up to conventions that
+the team has already agreed on.
+
+### Slicing
+
+I will use part of the proposed architecture from the [architecture document](./architecture.md)
+as an example.
+
+Let's imagine we have the following architectural concerns: some persistent (database) storage,
+some metrics, and communicating with some external servers (e.g. the 1inch API).
+
+Option 1:
+
+- `cowswap`
+    - `solution`
+        - `core` - contains the core logic
+        - `data` - contains the SQL queries, or alternatively implements some `Store` traits from `core`
+        - `metrics` - contains the details for metrics, or alternatively implements some `Metrics` traits from `core`
+        - `external` - implements external traits from `core`, e.g. `trait OneInch`, `trait Paraswap`, or similar.
+    - `quote`
+        - `core` - contains the core logic
+        - `data` - contains the SQL queries, or alternatively implements some `Store` traits from `core`
+        - `metrics` - contains the details for metrics, or alternatively implements some `Metrics` traits from `core`
+        - `external` - implements external traits from `core`, e.g. `trait EthBlockchain` for scanning transactions.
+
+Option 2:
+
+- `cowswap`
+    - `core` - contains all core logic
+        - `solution`
+        - `quote`
+    - `data` - contains all SQL queries, or alternatively implements all `Store` traits from `core` modules
+        - `solution`
+        - `quote`
+        - note: the above two could be a single module if there isn't too much code
+    - `metrics` - contains all details for metrics, or alternatively implements all `Metrics` traits from `core` modules
+        - `solution`
+        - `quote`
+        - note: the above two could be a single module if there isn't too much code
+    - `external` - implements all external traits from `core`
+        - `solution`
+        - `quote`
+        - note: the above two could be a single module if there isn't too much code
+
+Of course other than `cowswap` there could be other top-level modules. E.g. a top-level `data` module
+for defining some common DB types maybe.
+
+### The Conventions
+
+I would like the team to answer the following questions:
+
+- Do we want to mock the db?
```

Comment:
I think if a test actually involves the database mocking it could as much work as just using an in-memory database so we might as well use the real thing.